src/lava/magma/core/model/py/connection.py

```diff
@@ -0,0 +1,1702 @@
+    tx: np.ndarray = LavaPyType(np.ndarray, int, precision=6)
+    x1: np.ndarray = LavaPyType(np.ndarray, int, precision=7)
+    x2: np.ndarray = LavaPyType(np.ndarray, int, precision=7)
+
+    y0: np.ndarray = LavaPyType(np.ndarray, bool)
+    ty: np.ndarray = LavaPyType(np.ndarray, int, precision=6)
+    y1: np.ndarray = LavaPyType(np.ndarray, int, precision=7)
+    y2: np.ndarray = LavaPyType(np.ndarray, int, precision=7)
+    y3: np.ndarray = LavaPyType(np.ndarray, int, precision=7)
+
+    tag_2: np.ndarray = LavaPyType(np.ndarray, int, precision=6)
+    tag_1: np.ndarray = LavaPyType(np.ndarray, int, precision=8)
+
+    def __init__(self, proc_params: dict) -> None:
+        # If there is a plasticity rule attached to the corresponding process,
+        # add all necessary ports get access to all learning params
+        self._learning_rule = proc_params["learning_rule"]
+        self._shape = proc_params["shape"]
+
+        # Flag to determine whether weights have already been scaled.
+        self.weights_set = False
+
+        if self._learning_rule is not None:
+            # store shapes that useful throught the lifetime of this PM
+            self._store_shapes()
+            # store impulses and taus in ndarrays with the right shapes
+            self._store_impulses_and_taus()
+
+            # store active traces per dependency from learning_rule in ndarrays
+            # with the right shapes
+            self._build_active_traces_per_dependency()
+            # store active traces from learning_rule in ndarrays
+            # with the right shapes
+            self._build_active_traces()
+            # generate LearningRuleApplierBitApprox from ProductSeries
+            self._build_learning_rule_appliers()
+
+            # initialize TraceRandoms and ConnVarRandom
+            self._init_randoms()
+
+        super().__init__(proc_params)
+
+    def _set_wgts(self) -> np.ndarray:
+        wgt_vals = np.copy(self.weights)
+
+        # Saturate the weights according to the sign_mode:
+        # 0 : null
+        # 1 : mixed
+        # 2 : excitatory
+        # 3 : inhibitory
+        mixed_idx = np.equal(self.sign_mode, 1).astype(np.int32)
+        excitatory_idx = np.equal(self.sign_mode, 2).astype(np.int32)
+        inhibitory_idx = np.equal(self.sign_mode, 3).astype(np.int32)
+
+        min_wgt = -(2 ** 8) * (mixed_idx + inhibitory_idx)
+        max_wgt = (2 ** 8 - 1) * (mixed_idx + excitatory_idx)
+
+        saturated_wgts = np.clip(wgt_vals, min_wgt, max_wgt)
+
+        # Truncate least significant bits given sign_mode and num_wgt_bits.
+        num_truncate_bits = 8 - self.num_weight_bits + mixed_idx
+
+        truncated_wgts = np.left_shift(
+            np.right_shift(saturated_wgts, num_truncate_bits), num_truncate_bits
+        )
+
+        wgt_vals = truncated_wgts.astype(np.int32)
+        wgts_scaled = np.copy(wgt_vals)
+        self.weights_set = True
+        return wgts_scaled
```

Comment:
This is copied from the Dense Process. I factored this out in another PR, so once that is merged, you should be able to use that function directly. But generally, please do not copy code. Instead, try to find a way to refactor the design to make the code reusable. This is paramount to maintaining a sustainable, healthy code base.

Reply:
This was duplicated by accident. Thanks for spotting that.